src/fileops.c

```diff
@@ -415,29 +416,42 @@ int git_futils_mkdir_ext(
 			continue;
 
 		/* See what's going on with this path component */
-		opts->perfdata.stat_calls++;
+		mkdir_attempted = 0;
+		for (;;) {
+			opts->perfdata.stat_calls++;
+			if (p_lstat(make_path.ptr, &st) < 0) {
+				if (errno != ENOENT || mkdir_attempted == 1) {
+					giterr_set(GITERR_OS, "Failed to make directory '%s'", make_path.ptr);
+					error = GIT_EEXISTS;
+					goto done;
+				}
+
+				opts->perfdata.mkdir_calls++;
+				if (p_mkdir(make_path.ptr, mode) == 0) {
+					giterr_clear();
+					break;
+				} else if (errno != EEXIST) {
+					giterr_set(GITERR_OS, "Failed to make directory '%s'", make_path.ptr);
+					error = GIT_EEXISTS;
```

Comment:
This probably goes for the one above as well (though it looks like that was was there already), in these cases, we don't know exactly why it failed, it'd probably be better to return `-1` as we are handing the case in which the dir does exist inside this function, the caller wouldn't be expected to recover.